File: apps/mobile/lib/api/models/playlist.dart
import "package:hive/hive.dart";

part "playlist.g.dart";

@HiveType(typeId: 1)
class Playlist {
  @HiveField(0)
  final String playlistId;
  @HiveField(1)
  final String name;
  @HiveField(2)
  final String authorGoogleEmail;
  @HiveField(3)
  final String authorName;
  @HiveField(4)
  final String authorImage;
  @HiveField(5)
  final String? coverImage;
  @HiveField(6)
  final bool isPublic;

  const Playlist({
    required this.playlistId,
    required this.name,
    required this.authorGoogleEmail,
    required this.authorImage,
    required this.authorName,
    required this.coverImage,
    required this.isPublic,
  });

  factory Playlist.fromJson(Map<String, dynamic> json) {
    return Playlist(
      playlistId: json["playlistId"] as String,
      name: json["name"] as String,
      authorGoogleEmail: json["authorGoogleEmail"] as String,
      authorName: json["authorName"] as String,
      authorImage: json["authorImage"] as String,
      coverImage: json["coverImage"] as String?,
      isPublic: json["isPublic"] as bool,
    );
  }

  @override
  String toString() =>
      "Playlist(playlistId: $playlistId, name: $name, authorGoogleEmail: $authorGoogleEmail, authorImage: $authorImage, $authorName: authorName, coverImage: $coverImage, isPublic: $isPublic)";
}

File: apps/desktop/src/lib/schemas/playlist.ts
import { z } from "zod";

import { embeddedAlbumSchema, embeddedArtistSchema } from "./embedded";
import { musicTrackSchema } from "./music-track";

export const playlistSchema = z.object({
  playlistId: z.string(),
  name: z.string(),
  authorGoogleEmail: z.string(),
  authorName: z.string(),
  authorImage: z.string(),
  coverImage: z.string().nullish(),
  isPublic: z.boolean(),
});
export const playlistsSchema = z.object({
  playlists: z.array(playlistSchema).optional(),
});

export const playlistSongRecommendationsSchema = z.object({
  continuationToken: z.string(),
  tracks: z.array(musicTrackSchema),
});

export const playlistVideoTypeSchema = z.enum(["VIDEO_TYPE_TRACK", "VIDEO_TYPE_UVIDEO"]);

export const playlistTrackSchema = z.object({
  playlistId: z.string(),
  playlistTrackId: z.string(),
  title: z.string(),
  thumbnail: z.string(),
  positionInPlaylist: z.number(),
  isExplicit: z.boolean(),
  album: z.optional(embeddedAlbumSchema),
  artists: z.array(embeddedArtistSchema),
  duration: z.string(),
  videoId: z.string(),
  videoType: playlistVideoTypeSchema,
});
export const playlistTracksSchema = z.object({
  playlistTracks: z.array(playlistTrackSchema).optional(),
});
export const playlistTrackLikedStatusSchema = z.record(z.string(), z.boolean());
export const playlistTracksLikedStatusSchema = z.object({
  likedTracks: playlistTrackLikedStatusSchema,
});

export type PlaylistTrack = z.infer<typeof playlistTrackSchema>;
export type PlaylistTrackLikedStatus = z.infer<typeof playlistTrackLikedStatusSchema>;
export type PlaylistVideoType = z.infer<typeof playlistVideoTypeSchema>;
export type Playlist = z.infer<typeof playlistSchema>;
export type Playlists = z.infer<typeof playlistsSchema>;
export type PlaylistSongRecommendations = z.infer<typeof playlistSongRecommendationsSchema>;
